audittools/backing_store_sql.go

```diff
@@ -0,0 +1,316 @@
+		registry.MustRegister(s.writeCounter, s.readCounter, s.errorCounter, s.sizeGauge)
+	}
+}
+
+// Write implements BackingStore.
+func (s *SQLBackingStore) Write(event cadf.Event) error {
+	maxEvents := s.MaxEvents.UnwrapOr(10000)
+	count, err := s.countEvents()
+	if err != nil {
+		s.errorCounter.WithLabelValues("write_count").Inc()
+		return fmt.Errorf("audittools: failed to check event count: %w", err)
+	}
+
+	if count >= int64(maxEvents) {
+		s.errorCounter.WithLabelValues("write_full").Inc()
+		return fmt.Errorf("%w: current size %d exceeds limit %d", ErrBackingStoreFull, count, maxEvents)
+	}
+
+	eventData, err := json.Marshal(event)
+	if err != nil {
+		s.errorCounter.WithLabelValues("write_marshal").Inc()
+		return fmt.Errorf("audittools: failed to marshal event: %w", err)
+	}
+
+	// String concatenation safe after Init() table name validation (prevents SQL injection).
+	// Provides better performance than fmt.Sprintf for simple query construction.
+	//nolint:gosec // G202: Table name validated in Init() with regex matching PostgreSQL identifier rules
+	query := "INSERT INTO " + s.TableName + " (event_data) VALUES ($1)"
+	if _, err := s.db.Exec(query, eventData); err != nil {
+		s.errorCounter.WithLabelValues("write_insert").Inc()
+		return fmt.Errorf("audittools: failed to insert event: %w", err)
+	}
+
+	s.writeCounter.Inc()
+	return nil
+}
+
+// ReadBatch implements BackingStore.
+func (s *SQLBackingStore) ReadBatch() ([]cadf.Event, func() error, error) {
+	batchSize := s.BatchSize.UnwrapOr(100)
+	// String concatenation safe after Init() validation. ORDER BY uses index for efficient FIFO reads.
+	//nolint:gosec // G202: Table name validated in Init() with regex matching PostgreSQL identifier rules
+	query := "SELECT id, event_data FROM " + s.TableName + " ORDER BY created_at ASC, id ASC LIMIT $1"
+
+	rows, err := s.db.Query(query, batchSize)
+	if err != nil {
+		s.errorCounter.WithLabelValues("read_query").Inc()
+		return nil, nil, fmt.Errorf("audittools: failed to query events: %w", err)
+	}
+	defer rows.Close()
+
+	// Preallocate based on known batch size to avoid reallocations during iteration.
+	events := make([]cadf.Event, 0, batchSize)
+	eventIDs := make([]int64, 0, batchSize)
+
+	for rows.Next() {
+		var id int64
+		var eventData []byte
+
+		if err := rows.Scan(&id, &eventData); err != nil {
+			s.errorCounter.WithLabelValues("read_scan").Inc()
+			return nil, nil, fmt.Errorf("audittools: failed to scan event: %w", err)
+		}
+
+		var event cadf.Event
+		if err := json.Unmarshal(eventData, &event); err != nil {
+			s.errorCounter.WithLabelValues("read_unmarshal").Inc()
+			// Skip corrupted events rather than failing the entire batch - allows partial recovery.
+			continue
+		}
+
+		events = append(events, event)
+		eventIDs = append(eventIDs, id)
+	}
+
+	if err := rows.Err(); err != nil {
+		s.errorCounter.WithLabelValues("read_rows").Inc()
+		return nil, nil, fmt.Errorf("audittools: failed to iterate events: %w", err)
+	}
```

Comment:
Boilerplate can be replaced with https://pkg.go.dev/github.com/sapcc/go-bits/sqlext#ForeachRow.